### ISSUE TITLE
Add CSP policy

### ISSUE THREAD

**Issue by miketaylr** [webcompat/webcompat.com#763]
Probably something like:

```
Content-Security-Policy: default-src 'self'; img-src *; script-src 'self' www.google-analytics.com;
```

This allows images from anywhere, Google Analytics scripts to run, and everything else must come from webcompat.com. 

We need to probably fix a few things so we don't have to enable `'unsafe-inline'` (there are a few inline scripts in templates, but we could just as easily put them in external scripts).

We might just deploy report-only mode: http://www.w3.org/TR/CSP/#content-security-policy-report-only-header-field and let that run for a few weeks to see if we would break anything with that policy. We'd just need to make a POST endpoint to log errors.

**Comment by miketaylr** [comment]
Maybe "bad-next-patch" :smiling_imp: 

I don't actually know how to add a header to every request in Flask, so that would require a bit of research.

**Comment by karlcow** [comment]
add a header to every HTTP responses sent by Flask.

```python
app = Flask(__name__)

@app.after_request
def apply_csp(response):
    response.headers["Austin"] = "Mike is burger!"
    return response
```


Some documentation to do it **by view** with a decorator
http://flask.pocoo.org/snippets/56/


For remembering in the future:
It's also possible to subclass the response object
http://flask.readthedocs.org/en/latest/api/#response-objects

```python
>>> from flask import Flask
>>> webc = Flask(__name__)
>>> webc.make_response('Hello Mike')
<Response 10 bytes [200 OK]>
>>> webc.make_response('Hello Mike').headers
Headers([('Content-Type', u'text/html; charset=utf-8'), ('Content-Length', u'10')])
>>> from flask import Response
>>> class MyResponse(Response):
...     default_mimetype = 'text/plain'
... 
>>> class MyFlask(Flask):
...     response_class = MyResponse
... 
>>> webc = MyFlask(__name__)
>>> webc.make_response('Hello Mike').headers
Headers([('Content-Type', u'text/plain; charset=utf-8'), ('Content-Length', u'10')])
```

**Comment by miketaylr** [comment]
Cool, thanks @karlcow!

**Comment by miketaylr** [comment]
Other good default headers for security here: 

https://securityheaders.io/?q=webcompat.com&followRedirects=on

**Comment by miketaylr** [comment]
Opened #1329 to make the reporting endpoint, which should land before first.

**Comment by miketaylr** [comment]
(related, we went from an F to a B in the past week by adding various security headers, secure flag for session cookies, HSTS includeSubdomains, etc. -- fixing this bug should get us up to an A)

**Comment by miketaylr** [comment]
https://observatory.mozilla.org/analyze.html?host=webcompat.com

**Comment by karlcow** [comment]
also https://cspvalidator.org/

**Comment by miketaylr** [comment]
[Let's start with this recommendation](https://wiki.mozilla.org/Security/Guidelines/Web_Security):

> It is recommended to start with a reasonably locked down policy such as default-src 'none'; img-src 'self'; script-src 'self'; style-src 'self' and then add in sources as revealed during testing. 

Then come back in a few weeks and see what the report tells us. 

I'll land this report-only header in #1329 (and probably fix #1328 at the same time).

**Comment by miketaylr** [comment]
> also https://cspvalidator.org/

**Comment by miketaylr** [comment]
> 502 for now.

Back up, and our policy validates.

**Comment by miketaylr** [comment]
So, in 6 days we've had 13416 reports. Probably don't need to wait much longer to learn more. :)

Will try to summarize what I see in this issue, and we can go from there.

**Comment by miketaylr** [comment]
* 608 instances of `"script-sample":"call to eval() or related function blocked by CSP","source-file":"https://webcompat.com/js/webcompat.min.js?9f6030d22addbc1e768230fbf58d97a4","violated-directive":"script-src https://webcompat.com https://www.google-analytics.com"`.
  likely a lodash template. we can fix this by adding `unsafe-eval` to `script-src` policy, until we move to something like pre-compiled templates.
* 3337 instances of `"script-sample":"background-color`, `"violated-directive":"style-src https://webcompat.com"`. We have an inline style somewhere on the issues page. We can relax this with. `'unsafe-inline'`
* 1047 instances of `{"csp-report":{"blocked-uri":"https://www.google-analytics.com","document-uri":"https://webcompat.com/issues/new?url=http://www.coco.fr/chat/index.html","original-policy":"default-src 'none'; connect-src https://webcompat.com; font-src https://webcompat.com; img-src https://webcompat.com; script-src https://webcompat.com https://www.google-analytics.com; style-src https://webcompat.com; report-uri https://webcompat.com/csp-report","referrer":"","violated-directive":"img-src https://webcompat.com"}}`
  so it looks like google analytics is loading a tracking pixel or something. we need to whitelist that domain for `img-src`
* 1521 reports that contain `{"csp-report":{"document-uri":"https://webcompat.com/issues/751"` -- it looks like these show up in search engine results...and something happens. Dunno.
* 1647 reports like `{"csp-report":{"document-uri":"https://webcompat.com/issues/631","referrer":"https://www.google.es/","violated-directive":"img-src","effective-directive":"img-src","original-policy":"default-src 'none'; connect-src 'self'; font-src 'self'; img-src 'self'; script-src 'self' https://www.google-analytics.com; style-src 'self'; report-uri /csp-report","disposition":"report","blocked-uri":"https://avatars0.githubusercontent.com/u/1133538?v=3","status-code":0}}`, (or a different sub-domain) which means we need to whitelist the github avatars domains. We should add `*.githubusercontent.com` to `img-src`
* 311 reports about inline scripts. `"violated-directive":"script-src https://webcompat.com https://www.google-analytics.com` should allow `unsafe-inline` for now.

Let me make those changes, and we can look back after a few more days.

**Comment by miketaylr** [comment]
OK! 3 weeks later, let's look at what the report log has to teach us:

`{"csp-report":{"document-uri":"https://webcompat.com/tools/cssfixme","referrer":"https://github.com/webcompat/webcompat.com/issues/905","violated-directive":"font-src","effective-directive":"font-src","original-policy":"default-src 'none'; connect-src 'self'; font-src 'self'; img-src 'self' https://www.google-analytics.com https://*.githubusercontent.com; script-src 'self' 'unsafe-eval' 'unsafe-inline' https://www.google-analytics.com; style-src 'self' 'unsafe-inline'; report-uri /csp-report","disposition":"report","blocked-uri":"data","status-code":0}}`

I guess GA adds a data URI JS snippet... gross. I guess we allow this now and look to remove it later.

`{"csp-report":{"blocked-uri":"data","document-uri":"https://webcompat.com/issues/new?url=http%3A%2F%2Fqtipr.com%2F&src=desktop-reporter","original-policy":"default-src 'none'; connect-src https://webcompat.com; font-src https://webcompat.com; img-src https://webcompat.com https://www.google-analytics.com https://*.githubusercontent.com; script-src https://webcompat.com 'unsafe-eval' 'unsafe-inline' https://www.google-analytics.com; style-src https://webcompat.com 'unsafe-inline'; report-uri https://webcompat.com/csp-report","referrer":"","violated-directive":"img-src https://webcompat.com https://www.google-analytics.com https://*.githubusercontent.com"}}`

This is because we use data image URIs for upload previews. We should allow this.

`{"csp-report":{"document-uri":"https://webcompat.com/issues?page=1&per_page=50&state=open&stage=all&sort=created&direction=desc","referrer":"https://webcompat.com/","violated-directive":"font-src","effective-directive":"font-src","original-policy":"default-src 'none'; connect-src 'self'; font-src 'self'; img-src 'self' https://www.google-analytics.com https://*.githubusercontent.com; script-src 'self' 'unsafe-eval' 'unsafe-inline' https://www.google-analytics.com; style-src 'self' 'unsafe-inline'; report-uri /csp-report","disposition":"report","blocked-uri":"data","status-code":0}}`

I believe this is the data URL thing


Content Security Policy: The page’s settings observed the loading of a resource at https://api.github.com/search/issues?page=1&per_page=50&stage=all&sort=created&q=facebook.com+state%3Aopen+repo%3Awebcompat%2Fweb-bugs&order=desc (“connect-src https://webcompat.com”). A CSP report is being sent.  (unknown)

`{"csp-report":{"document-uri":"https://webcompat.com/issues?q=bancopopular","referrer":"https://webcompat.com/","violated-directive":"connect-src","effective-directive":"connect-src","original-policy":"default-src 'none'; connect-src 'self'; font-src 'self'; img-src 'self' https://www.google-analytics.com https://*.githubusercontent.com; script-src 'self' 'unsafe-eval' 'unsafe-inline' https://www.google-analytics.com; style-src 'self' 'unsafe-inline'; report-uri /csp-report","disposition":"report","blocked-uri":"https://api.github.com/search/issues?q=bancopopular+repo%3Awebcompat%2Fweb-bugs","line-number":9,"column-number":3868,"source-file":"https://webcompat.com/js/webcompat.min.js?9f6030d22addbc1e768230fbf58d97a4","status-code":0}}
`

We need to whitelist the GitHub search API for non-logged in search.

I'll file a follow up bug to make these tweaks.

**Comment by miketaylr** [comment]
Once #1519 is merged, I'll make a PR to send `Content-Security-Policy` instead of `Content-Security-Policy-Report-Only`. 🚀 🌝 

Then we can file bugs to get rid of the what requires us to need `"script-src 'self' 'unsafe-eval' 'unsafe-inline'`.

**Comment by karlcow** [comment]
@miketaylr soon 2 weeks? Anything weird on the server? Do we jump?

**Comment by miketaylr** [comment]
Yeah, I think we look good here. Here's the last 100 reports:

https://gist.github.com/miketaylr/a542e607daf5b7143f7459a87c907b35

Basically it looks like a) google translate stuff (probably from chrome or addon users??) b) googlebot stuff (I'm guessing web-bugs 751 has a ton of SEO juice since it's about whatsapp c) js injected by addons

I'll write a PR to flip the switch.

**Comment by karlcow** [comment]
Closed by #1563
